usaspending_api/references/v2/views/naics.py

```diff
@@ -0,0 +1,178 @@
+            if len(naic.code) < 6:
+                result = OrderedDict()
+                result["naics"] = naic.code
+                result["naics_description"] = naic.description
+                result["count"] = (
+                    NAICS.objects.annotate(text_len=Length("code"))
+                    .filter(code__startswith=naic.code, text_len=6)
+                    .count()
+                )
+            else:
+                result = OrderedDict()
+                result["naics"] = naic.code
+                result["naics_description"] = naic.description
+                result["count"] = 0
+            results.append(result)
+        return results
+
+    def _filter_search(self, naics_filter: dict) -> dict:
+        tier1_codes = set([])
+        tier2_codes = set([])
+        tier3_codes = set([])
+
+        tier3_naics = NAICS.objects.annotate(text_len=Length("code")).filter(**naics_filter, text_len=6)
+        for naic in tier3_naics:
+            tier3_codes.add(naic.code)
+            tier2_codes.add(naic.code[:4])
+            tier1_codes.add(naic.code[:2])
+        tier2_naics = NAICS.objects.annotate(text_len=Length("code")).filter(**naics_filter, text_len=4)
+
+        for naic in tier2_naics:
+            tier2_codes.add(naic.code)
+            tier1_codes.add(naic.code[:2])
+        extra_tier2_naics = NAICS.objects.annotate(text_len=Length("code")).filter(code__in=tier2_codes, text_len=4)
+        tier1_naics = NAICS.objects.annotate(text_len=Length("code")).filter(text_len=2, code__in=tier1_codes)
+
+        tier2 = set(list(tier2_naics))
+        ex2 = list(extra_tier2_naics)
+        for naic in ex2:
+            tier2.add(naic)
+        tier2_results = {}
+        for naic in tier2:
+            result = OrderedDict()
+            result["naics"] = naic.code
+            result["naics_description"] = naic.description
+            result["count"] = (
+                NAICS.objects.annotate(text_len=Length("code")).filter(code__startswith=naic.code, text_len=6).count()
+            )
+            result["children"] = []
+            tier2_results[naic.code] = result
+        for naic in tier3_naics:
+            result = OrderedDict()
+            result["naics"] = naic.code
+            result["naics_description"] = naic.description
+            result["count"] = 0
+            tier2_results[naic.code[:4]]["children"].append(result)
+        tier1_results = {}
+        for naic in tier1_naics:
+            result = OrderedDict()
+            result["naics"] = naic.code
+            result["naics_description"] = naic.description
+            result["count"] = (
+                NAICS.objects.annotate(text_len=Length("code")).filter(code__startswith=naic.code, text_len=6).count()
+            )
+            result["children"] = []
+            tier1_results[naic.code] = result
+        for key in tier2_results.keys():
+            tier1_results[key[:2]]["children"].append(tier2_results[key])
```

Comment:
If we didi the tier1 results before the tier 2 results, could we just do this part in the tier 2 step like how tier 3 links to tier two? It's strange that this linking happens differently in different steps.

Reply:
no, we would have to iterate through the children of each tier1 child to find the correct tier2 child for each tier3, which was the entire reason we spent all that time changing how the filter function works.